lib/Transforms/InsertGpuAllocs.cpp

```diff
@@ -0,0 +1,363 @@
+//===- InsertGpuAllocs.cpp - InsertGpuAllocs Pass  -------*- C++ -*-===//
+//
+// Copyright 2022 Intel Corporation
+// Part of the IMEX Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file converts the memref.allocs for device side to gpu.allocs to
+/// distinguish between host & device side memory allocations.
+///
+//===----------------------------------------------------------------------===//
+
+#include <imex/Transforms/Transforms.h>
+
+namespace imex {
+
+mlir::StringRef getAllocSharedAttrName() { return "gpu.alloc_shared"; }
+
+struct InsertGPUAllocs
+    : public mlir::PassWrapper<InsertGPUAllocs,
+                               mlir::OperationPass<mlir::func::FuncOp>> {
+
+  InsertGPUAllocs() = default;
+  InsertGPUAllocs(bool gpuDealloc) { useGpuDealloc = gpuDealloc; }
+  InsertGPUAllocs(const InsertGPUAllocs &pass) : PassWrapper(pass) {}
+
+  virtual void
+  getDependentDialects(mlir::DialectRegistry &registry) const override {
+    registry.insert<mlir::memref::MemRefDialect>();
+    registry.insert<mlir::gpu::GPUDialect>();
+    registry.insert<mlir::scf::SCFDialect>();
+    registry.insert<mlir::arith::ArithmeticDialect>();
+  }
+
+  void runOnOperation() override {
+    auto func = getOperation();
+    auto &funcBody = func.getBody();
+    if (funcBody.empty()) {
+      return;
+    } else if (!llvm::hasSingleElement(funcBody)) {
+      func.emitError("Function must have exactly one block");
+      signalPassFailure();
+      return;
+    }
+
+    struct AccessType {
+      bool hostRead = false;
+      bool hostWrite = false;
+      bool deviceRead = false;
+      bool deviceWrite = false;
+    };
+
+    llvm::SmallMapVector<mlir::Operation *, AccessType, 8> gpuBufferAllocs;
+    llvm::SmallMapVector<unsigned, AccessType, 8> gpuBufferParams;
+    llvm::SmallMapVector<mlir::Operation *, AccessType, 8>
+        gpuGetMemrefGlobalParams;
+    auto &aliases = getAnalysis<mlir::BufferViewFlowAnalysis>();
+
+    auto getMemref = [](mlir::Operation *op)
+        -> llvm::Optional<mlir::SmallVector<mlir::Value, 4>> {
+      if (auto load = mlir::dyn_cast<mlir::memref::LoadOp>(op)) {
+        return {{load.memref()}};
+      } else if (auto store = mlir::dyn_cast<mlir::memref::StoreOp>(op)) {
+        return {{store.memref()}};
+      } else if (auto call = mlir::dyn_cast<mlir::func::CallOp>(op)) {
+        mlir::SmallVector<mlir::Value, 4> ret;
+        for (auto arg : call.operands()) {
```

Comment:
because there might be a func.call whose operands would be memrefs. basically we are just looking for all the memref producers

Reply:
This is something you could add as a comment :)

Reply:
>>because there might be a func.call whose operands would be memrefs. basically, we are just looking for all the memref producers

I am a bit confused by this statement.  Store doesn't produce memref. There are so many other dialects/ops producing memrefs.  Why we only care load/store/call op? What exact analysis and transformation we do in this pass for these three ops?  

getMemRef() is a very generic name, and the implementation just takes three op categories.  I think it should be modified to be a narrower name exactly describing what it is.

Reply:
I guess this is to find the use of memref  (produced or consumed) within the device kernel, whose memory buffers are likely being prepared by outside the device kernel. That's why it looks for load/store. 

For call, the parameter may or may not be defined outside the device kernel. I guess they are collected and then verified later.